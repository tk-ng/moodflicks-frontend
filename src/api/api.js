import axios from "axios";

const baseURL = import.meta.env.VITE_BASE_URL || "http://localhost:3001/";

/** API Class.
 *
 * Static class tying together methods used to get/send to to the API.
 * There shouldn't be any frontend-specific stuff here, and there shouldn't
 * be any API-aware stuff elsewhere in the frontend.
 *
 */

export class Api {
	// the token for interactive with the API will be stored here.
	static token;

	static async request(url, data = {}, method = "get") {
		// console.debug("API Call:", url, data, method);

		const headers = { Authorization: `Bearer ${Api.token}` };
		const params = method === "get" ? data : {};

		try {
			return (
				await axios({
					url,
					method,
					baseURL,
					data,
					params,
					headers,
					allowAbsoluteUrls: false,
				})
			).data;
		} catch (err) {
			console.error("API Error:", err.response);
			let message = err.response.data.error.message;
			throw Array.isArray(message) ? message : [message];
		}
	}

	// Individual API routes

	static async getUser(username) {
		let res = await this.request(`/users/${username}`);
		return res.user;
	}

	static async saveChanges({ username, data }) {
		let res = await this.request(
			`/users/${username}`,
			{ username, data },
			"patch"
		);
		return res.user;
	}

	static async deleteUser(username) {
		let res = await this.request(`/users/${username}`, {}, "delete");
		return res.user;
	}

	static async login(data) {
		let res = await this.request("auth/token", data, "post");
		return res.token;
	}

	static async signup(data) {
		let res = await this.request("auth/register", data, "post");
		return res.token;
	}

	static async getTrendingMovies() {
		let res = await this.request("/movies");
		return res.movies;
	}

	static async searchMovie(keyword) {
		let res = await this.request(`/movies/search/?keyword=${keyword}`);
		return res.movies;
	}

	static async getMovieDetail(id) {
		let res = await this.request(`/movies/${id}`);
		return res.movie;
	}

	static async getMovieCredits(id) {
		let res = await this.request(`/movies/${id}/credits`);
		return res.credits;
	}

	static async getActivities() {
		let res = await this.request(`/moods/activities`);
		return res.activities;
	}

	static async searchMood(keyword) {
		let res = await this.request(`/moods/search/?keyword=${keyword}`);
		return res.moods;
	}

	static async getMoods() {
		let res = await this.request(`/moods`);
		return res.moods;
	}

	static async getMoodsByUser(username) {
		let res = await this.request(`/users/${username}/moods`);
		return res.moods;
	}

	static async getMood(id) {
		let res = await this.request(`/moods/${id}`);
		return res.mood;
	}

	static async getMoviesByMood(id) {
		let res = await this.request(`/moods/${id}/movies`);
		return res.movies;
	}

	static async getMoodsByMovie(id) {
		let res = await this.request(`/movies/${id}/moods`);
		return res.moods;
	}

	static async getSimilarMoods(id) {
		let res = await this.request(`/moods/${id}/similar`);
		return res.moods;
	}

	static async getRecommendedMoviesByMood(id) {
		let res = await this.request(`/moods/${id}/recommended`);
		return res.movies;
	}

	static async createMood({ mood, username }) {
		let res = await this.request(`/moods`, { mood, username }, "post");
		return res.mood;
	}

	// needs username to verify in backend
	static async saveMoodChanges({ id, name, username }) {
		let res = await this.request(`/moods/${id}`, { name, username }, "patch");
		return res.message;
	}

	// needs username to verify in backend
	static async addMovieToMood({ moodId, movieId, username }) {
		const movies = await this.getMoviesByMood(moodId);

		if (movies.length) {
			const movieIds = movies.map((m) => m.id);
			if (movieIds.includes(movieId)) return null;
		}

		let res = await this.request(
			`/moods/${moodId}/movies/${movieId}`,
			{ username },
			"post"
		);
		return res.movieMood;
	}

	// needs username to verify in backend
	static async removeMovieFromMood({ moodId, movieId, username }) {
		let res = await this.request(
			`/moods/${moodId}/movies/${movieId}`,
			{ username },
			"delete"
		);
		return res.message;
	}

	// needs username to verify in backend
	static async deleteMood({ id, username }) {
		let res = await this.request(`/moods/${id}`, { username }, "delete");
		return res.message;
	}

	static async voteMood({ moodId, voteType, username }) {
		let res = await this.request(
			`/moods/${moodId}/vote/${voteType}`,
			{ username },
			"post"
		);
		return res.message;
	}
}
