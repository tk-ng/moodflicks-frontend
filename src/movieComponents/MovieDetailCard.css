.movie-card .dropdown-menu {
	top: auto !important;
}

.movie-card {
	max-width: 1200px;
}
