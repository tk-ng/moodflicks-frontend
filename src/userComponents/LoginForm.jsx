import { useFields } from "../hooks/useFields";
import {
	Form,
	FormGroup,
	Label,
	Input,
	Button,
	Row,
	Col,
	Card,
	CardBody,
	Alert,
} from "reactstrap";
import { useNavigate } from "react-router";
import { toast } from "react-toastify";

export const LoginForm = ({ login }) => {
	const navigate = useNavigate();

	const [formData, handleChange, resetForm, formErrors, setFormErrors] =
		useFields({
			username: "",
			password: "",
		});

	const handleSubmit = async (evt) => {
		evt.preventDefault();
		try {
			await login(formData);
			resetForm();
			navigate("/");
			toast.success(`Welcome back, ${formData.username}!`);
		} catch (err) {
			setFormErrors(err);
		}
	};

	return (
		<div className="LoginForm">
			<Row>
				<Col
					className="container"
					md={{
						offset: 3,
						size: 6,
					}}
					sm="12"
					lg={{
						offset: 4,
						size: 4,
					}}
				>
					<h3 className="mb-3">Log In</h3>
					<Card>
						<CardBody>
							<Form onSubmit={handleSubmit}>
								<FormGroup>
									<Label for="username">Username</Label>
									<Input
										id="username"
										name="username"
										value={formData.username}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="password">Password</Label>
									<Input
										id="password"
										name="password"
										type="password"
										value={formData.password}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								{formErrors.length !== 0 && (
									<Alert color="danger">
										<p className="small mb-0">{formErrors}</p>
									</Alert>
								)}
								<div className="d-grid">
									<Button color="primary">Submit</Button>
								</div>
							</Form>
						</CardBody>
					</Card>
				</Col>
			</Row>
		</div>
	);
};
