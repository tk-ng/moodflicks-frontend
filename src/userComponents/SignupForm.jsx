import { useFields } from "../hooks/useFields";
import { useNavigate } from "react-router";
import {
	Form,
	FormGroup,
	Label,
	Input,
	Button,
	Row,
	Col,
	Card,
	CardBody,
	Alert,
} from "reactstrap";

export const SignupForm = ({ signup }) => {
	const naviagte = useNavigate();
	const [formData, handleChange, resetForm, formErrors, setFormErrors] =
		useFields({
			username: "",
			password: "",
			firstName: "",
			lastName: "",
			email: "",
		});

	const handleSubmit = async (evt) => {
		evt.preventDefault();
		try {
			await signup(formData);
			resetForm();
			naviagte("/");
		} catch (err) {
			setFormErrors(err);
		}
	};

	return (
		<div className="SignupForm">
			<Row>
				<Col
					className="container"
					md={{
						offset: 3,
						size: 6,
					}}
					sm="12"
					lg={{
						offset: 4,
						size: 4,
					}}
				>
					<h3 className="mb-3">Sign Up</h3>
					<Card>
						<CardBody>
							<Form onSubmit={handleSubmit}>
								<FormGroup>
									<Label for="username">Username</Label>
									<Input
										id="username"
										name="username"
										value={formData.username}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="password">Password</Label>
									<Input
										id="password"
										name="password"
										type="password"
										value={formData.password}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="firstName">First Name</Label>
									<Input
										id="firstName"
										name="firstName"
										value={formData.firstName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="lastName">Last Name</Label>
									<Input
										id="lastName"
										name="lastName"
										value={formData.lastName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="email">Email</Label>
									<Input
										id="email"
										name="email"
										type="email"
										value={formData.email}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								{formErrors.length !== 0 && (
									<Alert color="danger">
										<p className="small mb-0">{formErrors}</p>
									</Alert>
								)}
								<div className="d-grid">
									<Button color="primary">Submit</Button>
								</div>
							</Form>
						</CardBody>
					</Card>
				</Col>
			</Row>
		</div>
	);
};
