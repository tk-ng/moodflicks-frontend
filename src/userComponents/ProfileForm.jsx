import { useFields } from "../hooks/useFields";
import { useNavigate, useParams, Navigate } from "react-router";
import { Api } from "../api/api";
import { UserContext } from "./UserContext";
import { useContext, useState, useEffect } from "react";
import {
	Form,
	FormGroup,
	Label,
	Input,
	Button,
	Row,
	Col,
	Card,
	CardBody,
	Alert,
} from "reactstrap";
import { toast } from "react-toastify";

export const ProfileForm = () => {
	const { currUser, setCurrentUser } = useContext(UserContext);

	const params = useParams();
	const naviagte = useNavigate();
	const [profileUser, setProfileUser] = useState();

	const [isLoading, setIsLoading] = useState(true);
	const [notFound, setNotFound] = useState(false);

	const [
		formData,
		handleChange,
		resetForm,
		formErrors,
		setFormErrors,
		setFormData,
	] = useFields({
		firstName: "",
		lastName: "",
		email: "",
	});

	const handleSubmit = async (evt) => {
		evt.preventDefault();
		try {
			const updatedUser = await Api.saveChanges({
				username: profileUser.username,
				data: formData,
			});
			if (currUser.username === profileUser.username)
				setCurrentUser((currentUser) => ({
					...currentUser,
					data: updatedUser,
				}));
			resetForm();
			toast.success("Profile Updated Successfully");
			naviagte(`/users/${profileUser.username}`);
		} catch (err) {
			setFormErrors(err);
		}
	};

	async function getUserInfo(username) {
		try {
			setIsLoading(true);
			setNotFound(false);

			const userResult = await Api.getUser(username);
			setProfileUser(userResult);
			setFormData({
				firstName: userResult.firstName,
				lastName: userResult.lastName,
				email: userResult.email,
			});
		} catch (err) {
			console.error("User not found:", err);
			setNotFound(true);
		} finally {
			setIsLoading(false);
		}
	}

	useEffect(() => {
		getUserInfo(params.username);
	}, [params.username]);

	if (isLoading) return <p>Loading...</p>;
	if (notFound) {
		toast.warning(`username: ${params.username} could not be found.`);

		return <Navigate to="/" />;
	}

	if (!currUser.isAdmin && params.username !== currUser.username) {
		toast.warning(`You are not logged in as ${params.username}.`);
		return <Navigate to={`/users/${currUser.username}`} />;
	}

	return (
		<div className="ProfileForm">
			<Row>
				<Col
					className="container"
					md={{
						offset: 3,
						size: 6,
					}}
					sm="12"
					lg={{
						offset: 4,
						size: 4,
					}}
				>
					<h3 className="mb-3">Profile</h3>
					<Card>
						<CardBody>
							<Form onSubmit={handleSubmit}>
								<FormGroup>
									<Label for="username">Username</Label>
									<Input
										id="username"
										name="username"
										value={params.username}
										disabled
									/>
								</FormGroup>
								<FormGroup>
									<Label for="firstName">First Name</Label>
									<Input
										id="firstName"
										name="firstName"
										value={formData.firstName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="lastName">Last Name</Label>
									<Input
										id="lastName"
										name="lastName"
										value={formData.lastName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="email">Email</Label>
									<Input
										id="email"
										name="email"
										type="email"
										value={formData.email}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								{formErrors.length !== 0 && (
									<Alert color="danger">
										<p className="small mb-0">{formErrors}</p>
									</Alert>
								)}
								<div className="d-grid">
									<Button color="primary">Save Changes</Button>
								</div>
							</Form>
						</CardBody>
					</Card>
				</Col>
			</Row>
		</div>
	);
};
