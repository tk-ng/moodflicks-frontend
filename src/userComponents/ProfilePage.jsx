import { useContext, useEffect, useState } from "react";
import { UserContext } from "./UserContext";
import { Navigate, useParams } from "react-router";
import { Api } from "../api/api";
import { Button } from "reactstrap";
import { useNavigate } from "react-router";
import { MoodList } from "../moodComponents/MoodList";
import { toast } from "react-toastify";

export const ProfilePage = () => {
	const params = useParams();
	const navigate = useNavigate();
	const [user, setUser] = useState();
	const [moods, setMoods] = useState([]);
	const { currUser } = useContext(UserContext);

	const [isLoading, setIsLoading] = useState(true);
	const [notFound, setNotFound] = useState(false);

	useEffect(() => {
		async function getUserInfo(username) {
			try {
				setIsLoading(true);
				setNotFound(false);
				const [userResult, moodsResult] = await Promise.all([
					Api.getUser(username),
					Api.getMoodsByUser(username),
				]);
				setUser(userResult);
				setMoods(moodsResult);
			} catch (err) {
				console.error("User not found:", err);
				setNotFound(true);
			} finally {
				setIsLoading(false);
			}
		}

		getUserInfo(params.username);
	}, [params.username]);

	if (isLoading) return <p>Loading...</p>;
	if (notFound) {
		toast.warning(`username: ${params.username} could not be found.`);
		return <Navigate to="/" />;
	}

	return (
		<div className="text-light">
			<span className="display-6">{user.username}</span>
			{currUser && user.username === currUser.username && (
				<Button
					onClick={() => navigate(`/users/${user.username}/edit`)}
					className="m-3 "
				>
					Edit Profile
				</Button>
			)}
			<div>
				{moods.length ? (
					<>
						<h3> Mood-based movie lists, curated by {user.username}</h3>
						<MoodList moods={moods} hideCreatedBy={true} />
					</>
				) : (
					<span>{user.username} has not created any mood lists yet.</span>
				)}
			</div>
		</div>
	);
};
