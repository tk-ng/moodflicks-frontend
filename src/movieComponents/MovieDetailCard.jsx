import {
	Card,
	CardBody,
	CardTitle,
	CardText,
	CardImg,
	Row,
	Col,
	Dropdown,
	DropdownToggle,
	DropdownMenu,
	DropdownItem,
} from "reactstrap";
import { useState, useContext } from "react";
import { PosterBaseUrl } from "../config";
import "./MovieDetailCard.css";
import { Api } from "../api/api";
import { UserContext } from "../userComponents/UserContext";
import { useNavigate } from "react-router";
import { toast } from "react-toastify";

export const MovieDetailCard = ({ movie, credits }) => {
	const navigate = useNavigate();
	const { currUser, currUserMoods } = useContext(UserContext);
	const posterUrl = PosterBaseUrl + "w500" + movie.poster_path;

	const [dropdownOpen, setDropdownOpen] = useState(false);
	const toggle = () => setDropdownOpen((prevState) => !prevState);

	const handleClick = async (evt, mood) => {
		evt.preventDefault();
		const exist = await Api.addMovieToMood({
			moodId: mood.id,
			movieId: movie.id,
			username: mood.createdBy,
		});
		navigate(`/moods/${mood.id}`);
		exist
			? toast.success("Success")
			: toast.warn("Movie already exists in your list");
	};

	return (
		<Card className="m-3 bg-dark text-white movie-card">
			<Row className="g-0">
				<Col md="4">
					<CardImg
						src={posterUrl}
						alt={movie.title}
						className="img-fluid rounded-start"
					/>
				</Col>
				<Col md="8">
					<CardBody>
						<CardTitle tag="h3">{movie.title}</CardTitle>
						<div className="text-start">
							<CardText>Director: {credits.director}</CardText>
							<CardText>Main cast: {credits.cast.join(", ")}</CardText>
							<CardText>
								Genres: {movie.genres.map(({ name }) => name).join(", ")}
							</CardText>
							Description:
							<CardText>{movie.overview}</CardText>
							<CardText>
								<small>
									{movie.status === "Released" ? (
										<>Released: {movie.release_date}</>
									) : (
										<>Status: {movie.status}</>
									)}
								</small>
							</CardText>
						</div>
						{currUser && (
							<div className="mt-3">
								<Dropdown
									isOpen={dropdownOpen}
									toggle={toggle}
									direction="down"
									className="text-start"
								>
									<DropdownToggle caret color="danger">
										Add to my list
									</DropdownToggle>
									<DropdownMenu>
										<DropdownItem header>Pick a mood list</DropdownItem>
										{currUserMoods.map((m) => (
											<DropdownItem
												key={m.id}
												onClick={(evt) => handleClick(evt, m)}
											>
												{m.mood}
											</DropdownItem>
										))}
									</DropdownMenu>
								</Dropdown>
							</div>
						)}
					</CardBody>
				</Col>
			</Row>
		</Card>
	);
};
