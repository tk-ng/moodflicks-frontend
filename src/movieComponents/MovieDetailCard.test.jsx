import { render, screen, waitFor } from "@testing-library/react";
import { Api } from "../api/api";
import { MovieDetailCard } from "./MovieDetailCard";
import { MemoryRouter, useParams } from "react-router";
import { UserContext } from "../userComponents/UserContext";
import userEvent from "@testing-library/user-event";

vi.mock("../api/api", () => ({
	Api: {
		getMoods: vi.fn(),
		searchMood: vi.fn(),
		getMovieDetail: vi.fn(),
		getMovieCredits: vi.fn(),
		getMoodsByMovie: vi.fn(),
	},
}));

vi.mock("react-router", async function () {
	const mod = await vi.importActual("react-router");
	return {
		...mod,
		useParams: vi.fn(),
		useSearchParams: vi.fn(),
	};
});

const testMovie = {
	id: "testMovieId",
	title: "testMovieTitle",
	genres: [{ name: "testMovieGenre1" }],
	overview: "testMovieOverview",
	release_date: "testMovieReleaseDate",
	status: "testMovieStatus",
	poster_path: "testMoviePosterPath",
};

const testCredits = {
	director: "testMovieDirector",
	cast: ["cast1", "cast2", "cast3"],
};

const testCurrUserMoods = [
	{
		movies: ["testid1", "testid2", "testid3", "testid4"],
		count: 4,
		mood: "testmood-1",
		id: "8",
		createdBy: "testuser2",
		totalVotes: "2",
	},
	{
		movies: ["testid5", "testid6", "testid7", "testid8", "testid9"],
		count: 5,
		mood: "testmood-2",
		id: "11",
		createdBy: "testuser1",
		totalVotes: "0",
	},
];

Api.getMovieDetail.mockResolvedValue(testMovie);

describe("Render the MovieDetailCard component", function () {
	useParams.mockReturnValue({ id: "testMovieId" });
	it("mounts without crashing", async function () {
		const { container, asFragment } = render(
			<MemoryRouter>
				<UserContext.Provider
					value={{
						currUser: null,
						currUserMoods: [],
					}}
				>
					<MovieDetailCard movie={testMovie} credits={testCredits} />
				</UserContext.Provider>
			</MemoryRouter>
		);
		expect(container).toBeInTheDocument();
		expect(asFragment).toMatchSnapshot();
	});

	it("renders the proper movie's detail for logged in user", async function () {
		render(
			<MemoryRouter>
				<UserContext.Provider
					value={{
						currUser: { username: "testuser" },
						currUserMoods: testCurrUserMoods,
					}}
				>
					<MovieDetailCard movie={testMovie} credits={testCredits} />
				</UserContext.Provider>
			</MemoryRouter>
		);
		expect(
			screen.getByText(`Director: ${testCredits.director}`, { exact: false })
		).toBeInTheDocument();
		expect(
			screen.getByText(
				`Main cast: ${testCredits.cast[0]}, ${testCredits.cast[1]}, ${testCredits.cast[2]}`,
				{ exact: false }
			)
		).toBeInTheDocument();
		expect(
			screen.getByText(`Genres: ${testMovie.genres[0].name}`, { exact: false })
		).toBeInTheDocument();
		expect(screen.getByText(testMovie.overview)).toBeInTheDocument();

		// Logged in user should see the Add to my list button
		expect(screen.getByText("Add to my list")).toBeInTheDocument();
	});

	it("The Add to my list dropdown shows the user's moods", async function () {
		render(
			<MemoryRouter>
				<UserContext.Provider
					value={{
						currUser: { username: "testuser" },
						currUserMoods: testCurrUserMoods,
					}}
				>
					<MovieDetailCard movie={testMovie} credits={testCredits} />
				</UserContext.Provider>
			</MemoryRouter>
		);
		const user = userEvent.setup();
		const AddToBtn = screen.getByText("Add to my list");
		await waitFor(() => expect(AddToBtn).toBeInTheDocument());
		await user.click(AddToBtn);

		// The user's mood should be displayed
		expect(screen.getByText("Pick a mood list")).toBeVisible();
		testCurrUserMoods.forEach((mood) =>
			expect(screen.getByText(mood.mood)).toBeVisible()
		);
	});

	it("renders the proper movie's detail for logged out user", async function () {
		render(
			<MemoryRouter>
				<UserContext.Provider
					value={{
						currUser: null,
						currUserMoods: [],
					}}
				>
					<MovieDetailCard movie={testMovie} credits={testCredits} />
				</UserContext.Provider>
			</MemoryRouter>
		);
		expect(
			screen.getByText(`Director: ${testCredits.director}`, { exact: false })
		).toBeInTheDocument();
		expect(
			screen.getByText(
				`Main cast: ${testCredits.cast[0]}, ${testCredits.cast[1]}, ${testCredits.cast[2]}`,
				{ exact: false }
			)
		).toBeInTheDocument();
		expect(
			screen.getByText(`Genres: ${testMovie.genres[0].name}`, { exact: false })
		).toBeInTheDocument();
		expect(screen.getByText(testMovie.overview)).toBeInTheDocument();

		// Logged out user should NOT see the Add to my list button
		expect(screen.queryByText("Add to my list")).not.toBeInTheDocument();
	});
});
