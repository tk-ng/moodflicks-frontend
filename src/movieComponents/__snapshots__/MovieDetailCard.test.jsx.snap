// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render the MovieDetailCard component > mounts without crashing 1`] = `[Function]`;
